### PR TITLE
Encode mp4 and thumnail as frames are captured from ogre

### PR DESCRIPTION
**Pull request by colhom**
If you're running this outside the docker container, you'll need to have [xxHash](https://github.com/Cyan4973/xxHash) installed on your system.

```
git clone https://github.com/Cyan4973/xxHash
cd xxHash
make
sudo make install
sudo ldconfig
```

`rviz` now will output `captured.mp4`, `captured_keyed.mp4` and `thumnail.jpg` - no post-processing required. 

Highlights:
  * encodes both mp4s on the fly. frame-buffer level control from ogre screenshot to the mp4 container! 
  * uses xxHash to detect unchanged frames (for keyed mp4)
  * creates a thumnail halfway through the bag (using QImage jpeg export).

### CODE REVIEW

**Comment by IainBrookshaw** [comment]
I think something may be missing. I get build errors (installed `xxHash`).
The function `avcodec_parameters_from_context` does not seem to be defined anywhere.
I strongly suspect something isn't installed -- what defines `AVStream`?

```
/home/rvn-ros/Documents/RavenOps/rviz_ws/src/rviz/src/rviz_video_encoder/video_encoder.c: In function ‘video_encoder_init’:
/home/rvn-ros/Documents/RavenOps/rviz_ws/src/rviz/src/rviz_video_encoder/video_encoder.c:160:11: warning: implicit declaration of function ‘avcodec_parameters_from_context’ [-Wimplicit-function-declaration]
     ret = avcodec_parameters_from_context(enc->st->codecpar, enc->cc);
           ^
/home/rvn-ros/Documents/RavenOps/rviz_ws/src/rviz/src/rviz_video_encoder/video_encoder.c:160:50: error: ‘AVStream {aka struct AVStream}’ has no member named ‘codecpar’
     ret = avcodec_parameters_from_context(enc->st->codecpar, enc->cc);
                                                  ^
/home/rvn-ros/Documents/RavenOps/rviz_ws/src/rviz/src/rviz_video_encoder/video_encoder.c: In function ‘video_encoder_send_frame’:
/home/rvn-ros/Documents/RavenOps/rviz_ws/src/rviz/src/rviz_video_encoder/video_encoder.c:191:11: warning: implicit declaration of function ‘avcodec_send_frame’ [-Wimplicit-function-declaration]
     ret = avcodec_send_frame(enc->cc,frame);
           ^
/home/rvn-ros/Documents/RavenOps/rviz_ws/src/rviz/src/rviz_video_encoder/video_encoder.c:202:15: warning: implicit declaration of function ‘avcodec_receive_packet’ [-Wimplicit-function-declaration]
         ret = avcodec_receive_packet(enc->cc,&enc->pbuf);
               ^
rviz/src/rviz_video_encoder/CMakeFiles/rviz_video_encoder.dir/build.make:62: recipe for target 'rviz/src/rviz_video_encoder/CMakeFiles/rviz_video_encoder.dir/video_encoder.c.o' failed
make[2]: *** [rviz/src/rviz_video_encoder/CMakeFiles/rviz_video_encoder.dir/video_encoder.c.o] Error 1
CMakeFiles/Makefile2:3047: recipe for target 'rviz/src/rviz_video_encoder/CMakeFiles/rviz_video_encoder.dir/all' failed
make[1]: *** [rviz/src/rviz_video_encoder/CMakeFiles/rviz_video_encoder.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[  9%] Built target rviz_automoc
Makefile:138: recipe for target 'all' failed
```

**Comment by colhom** [comment]
@IainBrookshaw it turns out you'll need the `v4` libav libraries installed as well. My mistake for not mentioning this.

```sh
apt-get update
apt-get install software-properties-common
￼add-apt-repository ppa:jonathonf/ffmpeg-4
apt-get update
apt-get install libavformat-dev libavcodec-dev libavutil-dev  libswscale-dev libx264-155
```

Those deps will allow you to build and run.

**Comment by delaneyj** [comment]
Awesome news!  Will try in the morning

File: src/rviz/visualization_manager.cpp [hunk 1]
```diff
@@ -487,62 +512,134 @@ void VisualizationManager::onUpdate()
...
+          if (private_->venc_keyed_ == NULL){
+              ROS_ERROR("Failed to create capture keyed video encoder");
+              exit(EXIT_FAILURE);
+          }
+
+      }else{
+
+          // Ensure image dimensions have not changed
+          // RVN:TODO perhaps support this (if we have a use case)
+          if (screenshot_.width() != private_->venc_width_ || screenshot_.height() != private_->venc_height_)
```
Comment (IainBrookshaw) [inline src/rviz/visualization_manager.cpp:574]: This is currently being triggered:
`Detected change in screenshot dimensions: was -1493129544x1866242144, now is 1881x2088`
This is a result of venc_ not being initialized.

File: src/rviz/visualization_manager.cpp [hunk 2]
```diff
@@ -487,62 +512,134 @@ void VisualizationManager::onUpdate()
...
-      {
-        QString error_message;
-        if (writer.error() == QImageWriter::UnsupportedFormatError)
-        {
-          QString suffix = filename.section('.', -1);
-          QString formats_string;
-          QList<QByteArray> formats = QImageWriter::supportedImageFormats();
-          formats_string = formats[0];
-          for (int i = 1; i < formats.size(); i++)
+      if ( private_->venc_ == NULL){
```
Comment (IainBrookshaw) [inline src/rviz/visualization_manager.cpp:530]: This is never triggered -- `venc_` is not initialized to anything -- just points to uninitialized memory, so is not guaranteed to be null

File: src/rviz/visualization_manager.cpp [hunk 3]
```diff
@@ -121,6 +140,10 @@ class VisualizationManagerPrivate
   ros::NodeHandle update_nh_;
   ros::NodeHandle threaded_nh_;
   boost::mutex render_mutex_;
+  int venc_width_,venc_height_;
+  VideoEncoder *venc_,*venc_keyed_;
```
Comment (IainBrookshaw) [inline src/rviz/visualization_manager.cpp:144]: Should always be:
`VideoEncoder* venc_ = NULL;` and 
`VideoEncoder* venc_keyed_ = NULL;` as you are explicitly testing for `venc_ == NULL` later. There is no guarantee that this is going to be null if left uninitialized.

**Review by IainBrookshaw — CHANGES_REQUESTED** [review]
Minor problem with init of `venc_` pointer var. Just make it explicitly = NULL when created.
This is knock-on affects with window size checks which prevent rviz from running

Major problem, the mp4 output is scrambled:

broken_video

**Comment by colhom** [comment]
@IainBrookshaw have you pulled the latest: a40805fb3df57205030170db4afb8421b3ea1ce9 ?

File: src/rviz/visualization_manager.cpp (hunk 3 above)
Comment (colhom) [reply src/rviz/visualization_manager.cpp:144]: Ah, great catch. In my C++ noobery, mistakenly assumed that `new` initialized memory to zero.

**Comment by colhom** [comment]
@IainBrookshaw pointer initialization and % 4 frame dimension stuff has been addressed.

**Review by IainBrookshaw — APPROVED** [review]
as of commit a84f6786ed1cfc54f82a954c11072c2b11152c56 all parts seem to work perfectly!

File: src/rviz/visualization_manager.cpp (hunk 3 above)
Comment (delaneyj) [reply src/rviz/visualization_manager.cpp:144]: Why would you want that and not the random leftovers from previous applications?  I did this so many times in the initial implementation :)

**Pull request merged**